### PR TITLE
Hw04 Poshtarenko

### PR DESCRIPTION
**Pull request by EKCKABATOP54**

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+simplifyTests :: [(Expr ExprConstCaseType, Expr ExprConstCaseType)]
+simplifyTests = [ (Mult (Const 0) (Const 5), Const 0),
+          (Plus (Mult (Const 1) (Var "x")) (Const 0), Var "x"),
+          (Plus (Const 0) (       
+                Mult  (Const 1) (    
+                      Minus (Div (Const 42) (Const 1) ) (Const 0) 
+                )
+                )
+        , Const 42), 
+        (Plus (Const 1) (Mult (Const 0) (Var "x") ), Const 1)]
```
Comment (kajigor) [inline Main.hs:181]: It's also better to check that simplification doesn't happen when it is not supposed to happen.

File: Main.hs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        (Const 777, [],  Right 777),
+        (Pow (Div (Mult (Minus (Pow (Plus (SquareRoot(Const 25)) (Const 3)) (Const 2)) (Mult (Const 4) (Const 7))) (Const 2)) (Const 5)) (Const 3), [],  Right ( ((  ( ( sqrt 25 + 3)**2 - 4 * 7) * 2) / 5)**3)),  
+        let e1 = Mult (Const 4) (Const 5)
+            e2 = Plus (Const 1) (Const (-1)) in
+            (Div e1 e2, [],  Left (DivisionByZero e1 e2)),
+        let e = Minus (Const 1) (Const 100) in
+        (SquareRoot e, [],  Left (RootOfNegative e)),
+        (Var "x", [("x", 100)], Right 100),
+        (Plus (Var "x") (Var "y"), [("x", 10.0), ("y", 20.0)], Right 30.0),
+        (Plus (Var "z") (Var "y"), [("x", 10), ("y", 20)], Left $ UndefinedVariable "z")
```
[the remark this answers is not in the document]
Comment (EKCKABATOP54) [reply Main.hs]: I added a couple of tests, but i'm not sure i understand what you're talking about

[the remark this answers is not in the document]
Comment (kajigor) [reply Main.hs]: You still haven't found the other significant case for variables.

File: Main.hs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        (SquareRoot e, [],  Left (RootOfNegative e)),
+        (Var "x", [("x", 100)], Right 100),
+        (Plus (Var "x") (Var "y"), [("x", 10.0), ("y", 20.0)], Right 30.0),
+        (Plus (Var "z") (Var "y"), [("x", 10), ("y", 20)], Left $ UndefinedVariable "z"),
+        (Minus (Var "x") (Var "x"), [("x", 777)], Right 0), 
+        (Minus (Var "x") (Var "x"), [("x", 14)], Right 0), 
+        (Pow (Var "x") (Var "y"), [("x", 2), ("y", 10)], Right 1024 ),
+        (Pow (Var "x") (Var "y"), [("x", 2), ("y", 0)], Right 1 ),
+        (Pow (Var "x") (Var "y"), [("x", 0), ("y", 0)], Right 1 )
+        ]
```
Comment (kajigor) [inline Main.hs:111]: You don't need to re-test that multiplication, addition and so forth works. Focus on the variables and variable mappings.

**Comment by kajigor** [comment]
ok

**Pull request closed without merging**